### PR TITLE
ENH: Creating hdf5

### PR DESCRIPTION
**Pull request by EmmaRenauld**
## Description
Yeah!!!! The first really huge step is ready. Creation of the hdf5 dataset from the data organized as discussed in our lab meetings (such an eternity ago...).  Below is how to test exactly as I did. If you could test with your data to see if it is adapted to all of us it would be great!

## Testing data and script
Here is how you may test on Beluga. 

```
# Copy tractoflow output as symlinks
database_folder=./YOUR_PATH/hcp_1200/derivatives/
cp -as /home/USER/projects/rrg-descotea/datasets/hcp_1200/derivatives/tractoflow/output/results_ranByEmma/results/ $database_folder/preprocessed

# Create subject list
cd .$database_folder/preprocessed
echo s* > subjects_list.txt

# Organize data
bash organize_from_tractoflow.sh $database_folder $subject_list
bash organize_from_recobundles.sh $database_folder RecobundlesX/multi_bundles $subject_list

# Create hdf5
mask_for_standardization="masks/wm.nii.gz"
space='rasmm'
create_hdf5_dataset.py --bundles recobundlesX_SLF_2_L_m recobundlesX_SLF_2_R_m \
         --name $name --std_mask $mask_for_standardization \
         --space $space --logging debug $database_folder $config_file \
        $training_subjs $validation_subjs
```

where config_file is:
```
{
    "input": ["dwi/dwi_tractoflow.nii.gz", "anat/t1_tractoflow.nii.gz"]
}
```

## Have you
- [x ] Added a description of the content of this PR above
- [ ] Followed [proper commit message formatting](https://chris.beams.io/posts/git-commit/)
- [x ] Added data and a script to test this PR
- [x ] Made sure that PEP8 issues are resolved
- [x ] Tested the code yourself right before pushing
- [ ] Added unit tests to test the code you implemented

### CODE REVIEW

File: dwi_ml/data/creation/hdf5_creator.py [hunk 1]
```diff
@@ -1,586 +0,0 @@
...
-
-            # Transfer the streamlines to the reference space before bringing them
-            # to VOX space. NOTE: This is done in case the streamlines were tracked
-            # in a different space than the provided dataset reference
-            if output_tractogram is None:
-                output_streamlines_rasmm = []
-            else:
-                output_streamlines_rasmm = output_tractogram.streamlines
-            output_tractogram = StatefulTractogram(output_streamlines_rasmm,
-                                                   dwi_ref,
```
Comment (EmmaRenauld) [inline dwi_ml/data/creation/hdf5_creator.py:294]: @ppoulin91 , I am not sure I understood correctly this part. You were forcing the space but now it is an option for the user so I had to change this. Can you check that the management of space is correct in the new script? Thanks!

File: scripts_python/create_hdf5_dataset.py [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+            # Prepare subject folder for intermediate files
+            subj_intermediate_path = hdf5_dir.joinpath(subj_id)
+            if args.save_intermediate:
+                subj_intermediate_path.mkdir()
+
+            # Find subject's mask
+            if args.mask:
+                subj_mask_file = subj_input_dir.joinpath(args.mask)
+                subj_mask_data = nib.load(subj_mask_file)
+                subj_mask_data = subj_mask_data.get_fdata().astype(np.bool)
```
Comment (ppoulin91) [inline scripts_python/create_hdf5_dataset.py:195]: For a mask you should use `np.asanyarray(img.dataobj).astype(np.bool)`. This way you don't cast everything to float64 by default when calling `get_fdata()`

File: scripts_python/create_hdf5_dataset.py [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+                file_list = groups_config[group]
+                group_data, group_affine, group_header = process_group(
+                    group, file_list, args.save_intermediate, subj_input_dir,
+                    subj_intermediate_path, subj_mask_data)
+                logging.debug('...done. Now creating dataset from group.')
+                subj_hdf.create_dataset(group, data=group_data)
+                logging.debug('...done.')
+
+            # Reparing header. Should be arranged in Dipy 1.2.
+            group_header['srow_x'] = group_affine[0, :]
```
Comment (ppoulin91) [inline scripts_python/create_hdf5_dataset.py:210]: Why is this needed? I have never seen this requirement...

File: scripts_python/create_hdf5_dataset.py [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+            if streamlines is None:
+                logging.warning('Careful! Total tractogram for subject {} '
+                                'contained no streamlines!'.format(subj_id))
+            else:
+                streamlines_group = subj_hdf.create_group('streamlines')
+
+                streamlines_group.attrs['space_attributes'] = \
+                    str(tractogram.space_attributes)
+                # Accessing private Dipy values, but necessary
+                streamlines_group.create_dataset('data',
```
Comment (ppoulin91) [inline scripts_python/create_hdf5_dataset.py:242]: We don't know what space the user chose, this is dangerous...

File: dwi_ml/data/hdf5_creation.py [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+            group_data = np.append(group_data, data, axis=-1)
+        except ImportError:
+            raise ImportError('Data file {} could not be added to data group {}'
+                              '. Wrong dimensions?'.format(data_file, group))
+
+    # Save unnormalized data
+    if save_intermediate:
+        logging.debug('  Saving intermediate non standardized files.')
+        group_image = nib.Nifti1Image(group_data, group_affine)
+        output_fname = "{}_nonstandardized.nii.gz".format(group)
```
[the remark this answers is not in the document]
Comment (AntoineTheb) [reply dwi_ml/data/hdf5_creation.py:132]: C'est l'ensemble des data loadés appendés en un groupe, au lieu de tout séparé

File: dwi_ml/data/hdf5_creation.py [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+    bundles_dir : Path
+        Path to bundles folder.
+    bundles: List[str]
+        List of the bundles filenames to load.
+    header : nib.Nifti1Header
+        Reference used to load and send the streamlines in voxel space and to
+        create final merged SFT.
+    step_size: float
+        Step size to resample streamlines. If none, compress streamlines.
+    space: Space
```
[the remark this answers is not in the document]
Comment (AntoineTheb) [reply dwi_ml/data/hdf5_creation.py:168]: Effectivement il me semble que les SFTs, une fois sauvegardés, sont juste re-transférés vers un espace spécifique

File: dwi_ml/data/hdf5_creation.py [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+            logging.debug('  Resampling')
+            bundle = resample_streamlines_step_size(bundle, step_size)
+            logging.debug("Resampled streamlines' step size to {}mm"
+                          .format(step_size))
+        else:
+            logging.debug('  Compressing')
+            bundle = compress_sft(bundle)
+
+        # Compute euclidean lengths
+        output_lengths.extend(length(bundle.streamlines))
```
[the remark this answers is not in the document]
Comment (AntoineTheb) [reply dwi_ml/data/hdf5_creation.py:222]: Faudrait les envoyer en world space explicitement si on veux afficher la métrique plus bas

File: scripts_python/create_hdf5_dataset.py (hunk 3 above)
Comment (EmmaRenauld) [reply scripts_python/create_hdf5_dataset.py:210]: Hmmm.. I honestly don't remember, it's been too long. I think I was just printing everything as I was coding and I think I saw that it contained only zeros, or something. I surely didn't invent that it would be fixed in 1.2. Maybe I discussed it with François?

Do I keep it like that?

File: dwi_ml/data/hdf5_creation.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (EmmaRenauld) [reply dwi_ml/data/hdf5_creation.py:168]: Je suis un peu d'accord, mais quand j'ai écrit ça, je venais de gérer beaucoup de scripts où le space était une option. Je laisse, au cas où ça devraiendrait utile un jour.

File: scripts_python/create_hdf5_dataset.py (hunk 3 above)
Comment (ppoulin91) [reply scripts_python/create_hdf5_dataset.py:215]: I have no idea, I'll try to investigate a bit this week or next week

File: scripts_python/create_hdf5_dataset.py [hunk 8]
```diff
@@ -238,6 +241,8 @@ def _add_all_subjs_to_database(args, chosen_subjs: List[str],
 
                 streamlines_group.attrs['space_attributes'] = \
                     str(tractogram.space_attributes)
+                streamlines_group.attrs['space'] = space
```
Comment (ppoulin91) [inline scripts_python/create_hdf5_dataset.py:246]: Isn't this going to save only an `int`? I think we should save a string

Comment (EmmaRenauld) [reply scripts_python/create_hdf5_dataset.py:246]: I don't think it saves an int...

ipython:
```
from dipy.io.stateful_tractogram import Space
a = {}  
a['space'] = Space.RASMM
dict(a)                                                                                                                                                                       
```

Out: {'space': <Space.RASMM: 'rasmm'>}

Comment (ppoulin91) [reply scripts_python/create_hdf5_dataset.py:246]: You're right, sorry about that, I don't know why I was thinking about c++ :(

File: scripts_python/create_hdf5_dataset.py [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+                subj_hdf.create_dataset(group, data=group_data)
+                logging.debug('...done.')
+
+            # Reparing header. Should be arranged in Dipy 1.2.
+            group_header['srow_x'] = group_affine[0, :]
+            group_header['srow_y'] = group_affine[1, :]
+            group_header['srow_z'] = group_affine[2, :]
+
+            # Taking any group (the last) to save space information.
+            # All groups should technically have the same information.
```
Comment (ppoulin91) [inline scripts_python/create_hdf5_dataset.py:220]: I think this should be validated

File: scripts_python/create_hdf5_dataset.py (hunk 3 above)
Comment (ppoulin91) [reply scripts_python/create_hdf5_dataset.py:215]: Ok, so I checked quickly, and Dipy is not involved in fetching the header or affine (it is fetched straight from `nibabel.load().header`). This means that this "repair" is useful only if the header is already broken, in which case, it should be fixed elsewhere. 

Also, do we actually need the whole header? Some data might live in the same voxel space, but in a different world space (e.g. T1 vs. DWI), which means the header/affine might be a different, no?

File: dwi_ml/data/hdf5_creation.py [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+                         'file.'.format(data_file))
+
+    img = nib.load(data_file)
+    data = img.get_fdata(dtype=np.float32)
+    affine = img.affine
+    header = img.header  # Note. affine is header.get_sform().
+    img.uncache()
+
+    if len(data.shape) < 3:
+        raise NotImplementedError('Why would a data be less than 3D? We did '
```
Comment (AntoineTheb) [inline dwi_ml/data/hdf5_creation.py:71]: That error message might be a bit too "familiar", maybe something like " Data less than 3D is not handled" would be more appropriate.

File: dwi_ml/data/hdf5_creation.py [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+    for bundle_name in bundles:
+        # Find bundle name
+        # Note. glob uses str. Other possibility: Path.cwd().glob but creates
+        # a generator object.
+        logging.debug('    Loading bundle {}'.format(bundle_name))
+        bundle_possible_name = str(bundles_dir.joinpath('*' +
+                                                        bundle_name + '*'))
+        bundle_real_name = glob.glob(bundle_possible_name)
+        if len(bundle_real_name) == 0:
+            raise ValueError('Bundle {} not found!'.format(bundle_name))
```
Comment (AntoineTheb) [inline dwi_ml/data/hdf5_creation.py:201]: I don't get this check and the one below. What's the situation where these errors could be triggered ?

File: dwi_ml/data/hdf5_creation.py [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+        elif len(bundle_real_name) > 1:
+            raise ValueError(
+                'More than one file with name {}. Clean your bundles '
+                'folder.'.format(bundle_possible_name))
+
+        # Loading bundle and sending to wanted space
+        bundle = load_tractogram(bundle_real_name[0], header)
```
Comment (AntoineTheb) [inline dwi_ml/data/hdf5_creation.py:208]: I would actually add a check before for the extension to see if the filename is a tractogram

File: scripts_python/create_hdf5_dataset.py [hunk 13]
```diff
@@ -0,0 +1,262 @@
...
+    """Parse arguments, generate hdf5 dataset and save it on disk."""
+    args = _parse_args()
+
+    # Initialize logger
+    logging.basicConfig(level=str(args.logging).upper())
+
+    # Verify that dwi_ml_ready folder is found
+    dwi_ml_ready_folder = Path(args.database_folder + '/dwi_ml_ready')
+    if not dwi_ml_ready_folder.is_dir():
+        raise ValueError('The dwi_ml_ready folder was not found: '
```
Comment (AntoineTheb) [inline scripts_python/create_hdf5_dataset.py:99]: Change for "parser.error" ?

File: scripts_python/create_hdf5_dataset.py (hunk 3 above)
Comment (AntoineTheb) [reply scripts_python/create_hdf5_dataset.py:215]: Since the advent of the StatefulTractogram, is this still needed ? Header properties of nifitis and tractograms are weird and often misused and I wouldn't play with them too much unless strictly necessary.

File: scripts_python/create_hdf5_dataset.py (hunk 9 above)
Comment (AntoineTheb) [reply scripts_python/create_hdf5_dataset.py:220]: +1

**Review by AntoineTheb — COMMENTED** [review]
Haven't tested it but here's some comments

File: dwi_ml/data/hdf5_creation.py (hunk 11 above)
Comment (EmmaRenauld) [reply dwi_ml/data/hdf5_creation.py:201]: Yeah, I think I will remove that. 
At the time, I thought it was bad that your bundle list couldn't be, ex: OR_L (because recobundle's output is OR_L_m). I wanted to find automatically all bundles that have more or less the name you want. But now I think it's better not.

Comment (EmmaRenauld) [reply dwi_ml/data/hdf5_creation.py:201]: Do we allow a bundle to be not found, however? Should it be an option?

File: scripts_python/create_hdf5_dataset.py (hunk 13 above)
Comment (EmmaRenauld) [reply scripts_python/create_hdf5_dataset.py:99]: I don't know the standard here. If you think so. Others agree?

File: scripts_python/create_hdf5_dataset.py (hunk 3 above)
Comment (EmmaRenauld) [reply scripts_python/create_hdf5_dataset.py:215]: I will try to test this this week. I was already using stateful tractogram when I wrote the code. I will have to check why I had problems.

File: scripts_python/create_hdf5_dataset.py (hunk 13 above)
Comment (jhlegarreta) [reply scripts_python/create_hdf5_dataset.py:99]: I dislike answering a question with another one, but `ArgumentError` or `ArgumentTypeError` could also be considered. Not sure about the downsides/adequacy of these possibilities. But using Antoine's proposal for all parser-related errors might be a good default choice.

File: dwi_ml/data/hdf5_creation.py [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+
+    Returns
+    -------
+    output_tractogram : StatefulTractogram
+        All streamlines in voxel space.
+    output_lengths : List[float]
+        The euclidean length of each streamline
+    """
+    # Silencing SFT's logger if our logging is in DEBUG mode
+    set_sft_logger_level('WARNING')
```
Comment (EmmaRenauld) [inline dwi_ml/data/hdf5_creation.py:178]: I don't think I invented that. I must have copied it somewhere. Is that really useful?

Comment (AntoineTheb) [reply dwi_ml/data/hdf5_creation.py:178]: Well I wouldn't mind it. SFT's tend to produce *a lot* of output when in verbose mode.

Comment (EmmaRenauld) [reply dwi_ml/data/hdf5_creation.py:178]: Thanks!

File: scripts_python/create_hdf5_dataset.py [hunk 15]
```diff
@@ -0,0 +1,262 @@
...
+
+            if streamlines is None:
+                logging.warning('Careful! Total tractogram for subject {} '
+                                'contained no streamlines!'.format(subj_id))
+            else:
+                streamlines_group = subj_hdf.create_group('streamlines')
+
+                streamlines_group.attrs['space_attributes'] = \
+                    str(tractogram.space_attributes)
+                streamlines_group.attrs['space'] = space
```
Comment (ppoulin91) [inline scripts_python/create_hdf5_dataset.py:246]: Cette ligne là plante pour moi : `TypeError: Object dtype dtype('O') has no native HDF5 equivalent`
Je pense qu'il essaie de sauvegarder l'enum au lieu de la string elle même...

**Review by ppoulin91 — CHANGES_REQUESTED** [review]

File: scripts_python/create_hdf5_dataset.py (hunk 15 above)
Comment (ppoulin91) [reply scripts_python/create_hdf5_dataset.py:246]: Faut remplacer par `space.value` pour avoir la string `'rasmm'`; pour loader le space, il faut ensuite faire `Space(hdf_group["space"])`

**Comment by ppoulin91** [comment]
À part le bug de Space, tout fonctionne bien de mon côté!

File: scripts_python/create_hdf5_dataset.py (hunk 3 above)
Comment (EmmaRenauld) [reply scripts_python/create_hdf5_dataset.py:215]: > Also, do we actually need the whole header? Some data might live in the same voxel space, but in a different world space (e.g. T1 vs. DWI), which means the header/affine might be a different, no?

I checked with tractoflow output, and the affine from T1 and dwi are the same. Don't they always need to have the same if they are registered? I'm not sure.

However for the header, the dwi has:
```
'srow_x': array([-1. ,  0. ,  0. , 77.5], dtype=float32),
 'srow_y': array([   0. ,    1. ,    0. , -113.5], dtype=float32),
 'srow_z': array([  0.  ,   0.  ,   1.  , -63.25], dtype=float32),
```
Which is the same value as the affine, but the T1 has
```
 'srow_x': array([0., 0., 0., 0.], dtype=float32),
 'srow_y': array([0., 0., 0., 0.], dtype=float32),
 'srow_z': array([0., 0., 0., 0.], dtype=float32),
```
Which explains that I had issues. The step before registration is Crop_T1, and then the header fits with the affine. So possible solutions: 1) we don't care that headers don't fit, as long as affines fit. 2) We don't care about neither. 3) We keep this trick to "repair" the header.

What is your opinion?

File: scripts_python/create_hdf5_dataset.py (hunk 15 above)
Comment (EmmaRenauld) [reply scripts_python/create_hdf5_dataset.py:246]: Cool. Je pense que ça existait pas encore au début, et François m'avait fait caster le space manuellement, si je me souviens bien. Changé!

File: dwi_ml/data/hdf5_creation.py (hunk 11 above)
Comment (ppoulin91) [reply dwi_ml/data/hdf5_creation.py:200]: Missing bundles will definitely be a thing, we should support it

**Review by jhlegarreta — APPROVED** [review]
After doing a few tests, the scripts are working as described. The PR answers the current needs or use cases of those currently needing it, so merging it for the sake of going forward seems reasonable.

After a few discussions, any improvements should be better addressed in separate PRs.
